File: 733.flood-fill.rb
#
# @lc app=leetcode id=733 lang=ruby
#
# [733] Flood Fill
#

# @lc code=start
# @param {Integer[][]} image
# @param {Integer} sr
# @param {Integer} sc
# @param {Integer} color
# @return {Integer[][]}
VECTORS = [[0, 1], [0, -1], [1, 0], [-1, 0]]

def flood_fill(image, sr, sc, color)
  from_color = image[sr][sc]
  already_filled = Array.new(image.size) { Array.new(image[0].size, nil) }
  queue = []

  image[sr][sc] = color
  already_filled[sr][sc] = true
  queue.push([sr, sc])

  while queue.size > 0
    sr_and_sc = queue.pop
    current_sr, current_sc = sr_and_sc

    VECTORS.each do |vector|
      next_sr = current_sr + vector[0]
      next_sc = current_sc + vector[1]

      next unless 0 <= next_sr && next_sr < image.size
      next unless 0 <= next_sc && next_sc < image[0].size
      next unless image[next_sr][next_sc] == from_color
      next if already_filled[next_sr][next_sc]

      image[next_sr][next_sc] = color
      already_filled[next_sr][next_sc] = true
      queue.push([next_sr, next_sc])
    end
  end

  image
end
# @lc code=end

# p(flood_fill([[1, 1, 1], [1, 1, 0], [1, 0, 1]], 1, 1, 2))
# p(flood_fill([[0, 0, 0], [0, 0, 0]], 0, 0, 0))
p(flood_fill([[1, 1, 1], [1, 1, 0], [1, 0, 1]], 1, 1, 2))

File: 733.flood-fill.go
/*
 * @lc app=leetcode id=733 lang=golang
 *
 * [733] Flood Fill
 */

package main

// @lc code=start
type Queue[T any] []T

func (q *Queue[T]) Enqueue(v T) {
	*q = append(*q, v)
}

func (q *Queue[T]) Dequeue() T {
	v := (*q)[0]
	*q = (*q)[1:]
	return v
}

func (q *Queue[T]) IsEmpty() bool {
	return len(*q) == 0
}

type vector struct {
	dr int
	dc int
}

var vectors = [4]vector{
	{0, 1},
	{0, -1},
	{1, 0},
	{-1, 0},
}

func floodFill(image [][]int, sr int, sc int, color int) [][]int {
	origColor := image[sr][sc]

	visited := make([][]bool, len(image))
	for i := range image {
		visited[i] = make([]bool, len(image[i]))
	}

	type point struct {
		r int
		c int
	}
	var queue Queue[point]

	queue.Enqueue(point{sr, sc})

	for !queue.IsEmpty() {
		p := queue.Dequeue()

		if visited[p.r][p.c] {
			continue
		}
		if image[p.r][p.c] != origColor {
			continue
		}

		image[p.r][p.c] = color
		visited[p.r][p.c] = true

		for _, vec := range vectors {
			nextR := p.r + vec.dr
			nextC := p.c + vec.dc

			if 0 <= nextR && nextR < len(image) && 0 <= nextC && nextC < len(image[0]) {
				queue.Enqueue(point{nextR, nextC})
			}
		}
	}

	return image

}

// @lc code=end
